### ISSUE TITLE
XCTestExpectation fulfill() in Quick test causes NSInternalInconsistencyException

### ISSUE THREAD

**Issue by pcantrell** [Quick/Quick#330]
This code:

```
class QuickExpectationBug: QuickSpec {
    override func spec() {
        it("handles expectations correctly") {
            let expectation = self.expectationWithDescription("simple expectation")
            expectation.fulfill()
            self.waitForExpectationsWithTimeout(0.1, handler: nil)
        }
    }
}
```

…causes this exception:

```
Test Case '-[SiestaTests.QuickExpectationBug handles_expectations_correctly]' started.
2015-06-21 17:08:31.230 xctest[2044:34358] *** Assertion failure in -[XCTestExpectation fulfill], /Library/Caches/com.apple.xbs/Sources/XCTest_Sim/XCTest-8102.19/XCTestFramework/Classes/XCTestCase+AsynchronousTesting.m:399
/Users/paul/work/bustout/siesta/SiestaTests/QuickExpectationBug.swift:15: error: -[SiestaTests.QuickExpectationBug handles_expectations_correctly] : failed: caught "NSInternalInconsistencyException",
  "API violation - called -[XCTestExpectation fulfill] for simple expectation after
  the test case in which the expectation was created is no longer running.
  Created in test -[QuickExpectationBug (null)],
  current test is -[QuickExpectationBug handles_expectations_correctly]"
```

The same code works in an XCTestCase:

```
class ExpectationTest: XCTestCase {
    func testExpectation() {
        let expectation = self.expectationWithDescription("simple expectation")
        expectation.fulfill()
        self.waitForExpectationsWithTimeout(0.1, handler: nil)
    }
}
```

Something to do with Quick’s dynamically created test cases?

Using XC7.0 beta (7A120f), swift-2.0 branch of Quick.

**Comment by pcantrell** [comment]
This appears to be caused by the fact that +[QuickSpec initialize] builds all the Example instances in one go, using a shared dummy instance of the spec:

    + (void)initialize {
        ...
        QuickSpec *spec = [self new];

        @try {
            [spec spec];   // ← All it() calls happen here
        }
        @catch (NSException *exception) {
          ...
        }
    }

Any it() closures that capture self are capturing that dummy instance.

Possible solutions:

**Option 1**: Re-run the spec method before each test, so that self captures the correct instance of QuickSpec, then discard all Example instances except the relevant one.

  * Pros: Transparent. Makes the easy thing the correct thing.
  * Cons: May cause unintended side effects if describe() / context() closures do inline setup work. Possible performance hit for large suites.

**Option 2**: Allow it() closures to have the either the signature `QuickSpec -> ()` or the currently accepted `() -> ()`. If an example uses that new flavor, the closure is passed the correct QuickSpec instance, and code uses this instead of self:

    it("handles expectations correctly") { spec in
        let expectation = spec.expectationWithDescription("simple expectation")
        expectation.fulfill()
        spec.waitForExpectationsWithTimeout(0.1, handler: nil)
    }

**Option 3**: Provide a global method that returns the currently executing spec:

    it("handles expectations correctly") {
        let expectation = QuickSpec.currentSpec().expectationWithDescription("simple expectation")
        expectation.fulfill()
        QuickSpec.currentSpec().waitForExpectationsWithTimeout(0.1, handler: nil)
    }

I’m inclined to prefer option 2 if it doesn’t get too awkward with Swift’s finicky type checking, or option 3 if option 2 gets messy. @modocache, others, your thoughts?

Is there a version of option 3 somewhere that I’m missing? Seems like a thing that might already exist.

**Comment by jeffh** [comment]
I think there's some thought that needs to go into this.

Personally, I would rather split `QuickSpec` into 2 classes. One that is defines the `func spec()` (the tester api) and the other that integrates with XCTest (running infrastructure). The tradeoff is that splitting out `XCTestCase` from `QuickSpec` is definitely not pretty in any shape or form (reference: [Cedar's CDRSpec and XCTestCase separation](https://github.com/pivotal/cedar/blob/master/Source/iPhone/XCTest/CDRSpec+XCTestSupport.m))

But this separation might be more useful when the Linux-Swift compiler is released. I'm not optimistic in that XCTest (thus, Foundation) gets ported to Linux too.

This definitely leaves the XCTest assertion mechanism needing more thought. Maybe some adapter-like API that Quick exposes that simply defers to XCTest during execution.

Internal to Quick, I've just merged @modocache's [implementation of option 3](https://github.com/Quick/Quick/blob/master/Quick/QuickSpec.m#L7) on master. But I rather not have that as the preferred solution if possible.

Just my 2 cents.

**Comment by pcantrell** [comment]
Splitting QuickSpec into two classes as you describe seems like it could be a good idea, but it’s not something that needs to be coupled to fixing this problem.

That split would already be a breaking change with the current Quick API, so you're going to impact a few existing specs regardless. How much _more_ breakage would there be implementing the options above?

 * **Option 1:** A lot more breakage. XCTestCase methods would no longer be present on self in it() closures.
 * **Option 2:** No additional breakage. The spec passed to the closures is still valid, so specs with the new base class work as is.
 * **Option 3:** Modest additional breakage: QuickSpec.currentSpec() would be deprecated in favor of SomeNewQuickClassForXCTestCaseAccess.currentSpec().

Bottom line: I don’t think it’s worth holding up a fix for this issue unless you want to pursue option 1.

I still like option 2 best from the users’ point of view, though the internals are a bit messier. I can run through the implementation of that and make a PR for comparison if you want.

**Comment by jeffh** [comment]
It's not the amount of breakage I was concerned about, but the amount of times we break users. Since the coupling to XCTest is an implementation that leaks through, any migration off it will be considered a breakage.

Introducing a more explicit API to access XCTest currently just makes two places to break, an implicit expectation that `QuickSpec` is inheriting from `XCTestCase` and whatever API we choose to build out. FWIW, `QuickSpec` may just support the XCTestCase API for non-XCTestCase uses (just a random thought).

I'll let more @Quick/contributors pitch in thoughts than just me, but in regards to my thoughts about the options:

- I rather avoid **Option 1**. The performance hit is something I consider unacceptable for any large test suite based on working on other testing libraries.
- **Option 2** is proposed as an integration point for features, but they don't overlap well IMO. It smells like something needs a redesign pass when another features gets introduced using this mechanism. The current features using pass-to-closure style:
   - [https://github.com/Quick/Quick/issues/153] Shared Examples (I do like this idea though)
   - [https://github.com/Quick/Quick/issues/148 and https://github.com/Quick/Quick/pull/120] Metadata
- **Option 3** brings global state which could be avoided.

As I think through writing this out, I'd probably be ok accepting **Option 3** with it as an instance method instead of a class/static global.

Another idea would be to use that instance method to internally dispatch to the true XCTestCase that XCTest is running.

And just to be clear @pcantrell, thanks for your write ups and PR. It's a non-trivial amount of work you're putting in :clap:.

**Comment by pcantrell** [comment]
Yes, the amount of breakage for users (i.e. changing public contract, explicit or implicit) is what I was thinking of too.

I agree about the potential problems of **Option 1**. It is possible to make the actual traversal very fast, almost no overhead at all. (In a nutshell, remember the index sequence that leads to each it() block, e.g. [3,2,0,7], then do a pruned retraversal for each it() when running tests. Fast!) However, there's no way to prevent the performance nightmare of any non-lazy shared setup in context()/describe() blocks. Instant mess.

The other PRs you mention that potentially conflict with **Option 2** suggests making a `SpecContext` object that’s passed to each it(), which would let you add new features without breaking existing ones:

    it("handles expectations correctly") { context in
        let expectation = context.xcTestCase.expectationWithDescription(
            "simple expectation for \(context.metadata.specNameOrWhatever)")
        expectation.fulfill()
        context.xcTestCase.waitForExpectationsWithTimeout(0.1, handler: nil)
    }

Agreed that the global state of **Option 3** doesn’t smell quite right. XCTestCase is internally already built around that approach, though, it seems; we’re not going out on some crazy limb with it.

Your idea of making QuickSpec provide proxy implementations of all the standard XCTestCase methods is interesting. Call that **Option 4**. Two problems I see:

* Quick constantly plays catch-up to Apple adding methods to XCTestCase. (Not a big deal.)
* Would this force people to litter their code with `[weak self]` in order to prevent retain cycles? (That is a big deal; may also affect Option 1.)

<quote> thanks for your write ups and PR. It's a non-trivial amount of work you're putting in</quote>

Happy to help.

**Comment by jeffh** [comment]
Just because someone else does something crazy doesn't mean we have to.

The constant catch up doesn't seem to much of a burden, but the retain cycles could be cleared by having Quick explicitly nil blocks after their use. This only applies to `beforeEach`, `afterEach`, and `it` closures.

**Comment by pcantrell** [comment]
Re Option 3: It's not just a matter of “look, they used that pattern too;” it's that a global currentSpec() method doesn’t introduce any global state that doesn't already exist. At that point, I'm not sure what actual problem you're solving by doing contortions to avoid it. Still not elegant, but not a showstopper.

Re Option 4: The retain cycle is established as soon as the test method is defined (in `addInstanceMethodForExample`). This means two things: (1) Fixing the retain cycles by nilling out the closures after they execute means that all the test methods would be “run once only” — and that’s “run once” across all instances of the spec! (2) Retain cycles would still exist for any it() blocks that don't get executed, making focused execution a memory hazard. Both of those things seem dangerous. Fixing them would require some pretty substantial re-plumbing.

What do you think of the `SpecContext` idea above? It’s clean, and solves all the problems at hand: fixes this issue, no global state, no retain cycles, allows the QuickSpec/XCTestCase split, leaves room for the other PRs that also want to pass info via closure params, and leaves room for other future expansion.
